src/librustdoc/html/render.rs

```diff
@@ -1118,9 +1121,11 @@ impl DocFolder for Cache {
         let parent_pushed = match item.inner {
             clean::TraitItem(..) | clean::EnumItem(..) | clean::StructItem(..) => {
                 self.parent_stack.push(item.def_id);
+                self.parent_is_trait_impl = false;
                 true
             }
             clean::ImplItem(ref i) => {
+                self.parent_is_trait_impl = i.trait_.is_some();
```

Comment:
I think you can have an impl-in-an-impl, so this flag may need to be preserved and restored across the recursive calls?